Add GPU acceleration for gamma component transfer filters

nullhook opened the pull request:
#### 8a194767935ccd544f80cc447e29cd5c1d28c035
<pre>
Add GPU acceleration for gamma component transfer filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=296882">https://bugs.webkit.org/show_bug.cgi?id=296882</a>
<a href="https://rdar.apple.com/problem/157477030">rdar://problem/157477030</a>

Reviewed by Said Abou-Hallawa and Simon Fraser.

Support gamma-type feComponentTransfer in Core Image path. Previously,
only linear transfers were hardware-accelerated, forcing gamma
transforms to use slower software rendering.

Source/WTF/wtf/PlatformHave.h:
Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.h:
Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
(WebCore::isNullOr):
(WebCore::allChannelsMatch):
(WebCore::gammaKernel):
(WebCore::FEComponentTransferCoreImageApplier::supportsCoreImageRendering): Deleted.
(WebCore::FEComponentTransferCoreImageApplier::apply const): Deleted.
Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:

Canonical link: <a href="https://commits.webkit.org/298602@main">https://commits.webkit.org/298602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ef8c4324ea1dcb09dbade9a0e078f3579b8a0d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26184 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122042 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66548 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/05d5cf1e-c6e4-429b-a532-b75a873802c5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117882 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44234 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88112 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2e4bc08d-28a1-4e11-b4c0-19fc66ca588a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28994 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104081 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68522 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28115 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22190 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65712 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108102 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22325 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125195 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114521 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42880 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32186 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96866 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43246 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96651 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24596 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41911 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19787 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42769 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48366 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143205 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42236 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36918 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45575 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43945 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +68,49 @@
+
+    if (needsGamma) {
+        NSError *error = nil;
+        NSArray<CIKernel *> *kernels = [CIKernel kernelsWithMetalString:@R"(
```
we could cache `kernels` either in a member variable or static variable

nullhook replied:
cached

mwyrzykowski commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +68,49 @@
+    if (needsGamma) {
+        NSError *error = nil;
+        NSArray<CIKernel *> *kernels = [CIKernel kernelsWithMetalString:@R"(
+            [[stitchable]] float4 gammaFilter(coreimage::sampler src, float ampR, float expR, float offR,
```
is `[[stitchable]]` required here?

nullhook replied:
```
Kernel creation failed: Error Domain=CIKernel Code=12 "(null)" UserInfo={CINonLocalizedDescriptionKey=Cannot find a valid stitchable Metal function in the source}
```
yes, it seems so. maybe compiling with MTLLibrary data will make it unnecessary?

nullhook replied:
if we compile the kernel during build-time we can either go with `[[stitchable]]` or `extern "C"`. also, `[[stitchable]]` seems to be recommended.

mwyrzykowski replied:
That is fine then, thanks for confirming

mwyrzykowski commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +68,49 @@
+        })" error:&error];
+
+        if (error || !kernels || !kernels.count) {
+            WTFLogAlways("Kernel creation failed: %@", error);
```
since we are in WebCore code we could use `RELEASE_LOG`

mwyrzykowski commented on Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm:
```diff
@@ -0,0 +1,68 @@
+    if (!inputImage)
+        return false;
+
+    auto alphaFilter = [CIFilter filterWithName:@"CIColorMatrix"];
```
is there an advantage to using `CIColorMatrix` over `CAColorMatrix` here?

nullhook replied:
i think they behave similarly. it's just that CI filters only work with CIFilter subclasses. CAColorMatrix isn’t compatible here because CA filters are designed for CALayer, not CI.

mwyrzykowski reviewed (commented):
Since this is in WebCore, I think `RetainPtr` and `adoptNS(...)` are needed

mwyrzykowski reviewed (commented):
This looks good but would be good to get @shallawa's opinion as well

nullhook commented on Source/WebCore/platform/mac/VideoPresentationInterfaceMac.mm:
```diff
@@ -45,7 +45,10 @@
+#if !defined(SOFTLINK_AVKIT_FRAMEWORK_DEFINED)
+#define SOFTLINK_AVKIT_FRAMEWORK_DEFINED
 SOFTLINK_AVKIT_FRAMEWORK()
+#endif
```
i added a file in SourcesCocoa.txt which broke the unified build, so had to add these guards. fwiw, adding `@no-unify` caused a strange compiler error

@jyavenard

jyavenard replied:
I've had this error before , you can only use SOFTLINK_AVKIT_FRAMEWORK in a file once otherwise you'll get an error about the AVKit symbol being redefined .
you should instead use `SOFT_LINK_FRAMEWORK_FOR_SOURCE` here.

nullhook replied:
the issue is after adding @no-unify i get this error and `PictureInPictureSupport.h` seems like the relevant header. however, i can't find it's implementation

```
ld: warning: ignoring duplicate libraries: '-lobjc', '-lswiftCore'
Undefined symbols for architecture arm64e:
  "WebCore::supportsPictureInPicture()", referenced from:
      WebCore::DocumentPictureInPicture::pictureInPictureEnabled(WebCore::Document&) in UnifiedSource73.o
ld: symbol(s) not found for architecture arm64e
```

nullhook replied:
fixed this by adding an entry to build phase list in xcode

jyavenard commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,52 @@
 bool FEComponentTransferCoreImageApplier::supportsCoreImageRendering(const FEComponentTransfer& effect)
 {
-    auto isNullOrLinear = [] (const ComponentTransferFunction& function) {
+    auto isSupported = [] (const ComponentTransferFunction& function) {
```
no space between [] and ( as per code style.
```suggested
    auto isSupported = [](const ComponentTransferFunction& function) {
````

weinig commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,52 @@
+    std::call_once(onceFlag, [] {
+        NSError *error = nil;
+        NSArray<CIKernel *> *kernels = [CIKernel kernelsWithMetalString:@R"( /* NOLINT */
+            [[stitchable]] float4 gammaFilter(coreimage::sampler src, float ampR, float expR, float offR,
```
I'm not sure what the status quo on such things are, but can be this be compiled at build time and bundled with the framework as a binary, rather than being runtime constructed?

mwyrzykowski replied:
Yes, we could certainly do that. The cost is quite minimal given this shader's size, but at least we should file a bugzilla to make the build changes to load from a metallib.

nullhook replied:
yes, that's ideal. perhaps we can address it as a follow-up, since there may be more shaders coming

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,10 +104,37 @@
+        || m_effect->greenFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->blueFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA;
+
+    if (needsGamma) {
```
I think this block can be moved to a separate function.

shallawa commented on Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.h:
```diff
@@ -0,0 +1,50 @@
+
+namespace WebCore {
+
+class SourceAlphaCoreImageApplier final : public FilterEffectConcreteApplier<SourceAlpha> {
```
I think this change is not related to this PR. Can we move it to a separate patch?

nullhook replied:
fwiw, i think this change is important for this or any accelerated path because if any FilterEffect operates in software mode, the entire filter chain falls back to software rendering.

nullhook replied:
the above behavior i'm describing effectively negates the benefit of enabling "Accelerated Filter Rendering."

shallawa replied:
I understand. But this change is not necessary to get this PR compile and link. right? You can even submit a separate PR for the `SourceAlphaCoreImageApplier` only and land it before this patch if you prefer.

nullhook replied:
created a separate PR https://github.com/WebKit/WebKit/pull/49027

shallawa reviewed (commented):
Can't we have a layout test which turns the feature flag `AcceleratedFiltersEnabled` on and apply a simple `feComponentTransfer` to solid rectangle?

shallawa commented on Source/WebCore/SourcesCocoa.txt:
```diff
@@ -631,14 +632,14 @@ platform/mac/ThreadCheck.mm @no-unify
 platform/mac/UserAgentMac.mm
 platform/mac/ValidationBubbleMac.mm
-platform/mac/VideoPresentationInterfaceMac.mm
+platform/mac/VideoPresentationInterfaceMac.mm @no-unify
```
What is the reason for this change? Did you hit a compilation error with this source in the unified file? Please try to fix the compilation error within the unified file if possible.

nullhook replied:
https://github.com/WebKit/WebKit/pull/48898#discussion_r2252808916

nullhook replied:
adding `../SourceAlphaCoreImageApplier.mm` broke unified build in the files i've annotated as `@no-unify` — tried fixing compilation error within unified file which yielded into more errors.

shallawa replied:
Usually we use `@no-unify` when a source for sure can't be compiled within a unified file. For example the source defines these global
```
const static int x = 0;
// or
#define x foo()
````
But another source in the unified file defines
```
const static int x = 1;
// or
#define x bar()
```
Other than that we should try to fix the compilation errors within the unified file.

mwyrzykowski replied:
@shallawa is correct, we should add the missing transitive includes which is usually why unified builds break after adding new files.

C++ code should not rely on transitive includes.

nullhook replied:
this has been addressed: https://github.com/WebKit/WebKit/pull/49027

shallawa commented on Source/WebCore/SourcesCocoa.txt:
```diff
@@ -631,14 +632,14 @@ platform/mac/ThreadCheck.mm @no-unify
 platform/mac/WebCoreNSURLExtras.mm
 platform/mac/WebCoreObjCExtras.mm
-platform/mac/WebPlaybackControlsManager.mm
+platform/mac/WebPlaybackControlsManager.mm @no-unify
```
Ditto.

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +108,15 @@
+#if HAVE(CI_KERNEL)
+    bool needsGamma = m_effect->redFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->greenFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->blueFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA;
```
I think this is wrong. To call `applyGamma()`, I think all the channels have to be of type `FECOMPONENTTRANSFER_TYPE_GAMMA` or `FECOMPONENTTRANSFER_TYPE_UNKNOWN`. `ComponentTransferFunction` is like a union. The member type `type` decides which members of this structure are valid.

nullhook replied:
can you indicate this in the spec? i can't seem to find it.

shallawa replied:
This is not in the spec. This is how `applyLinear()` was implemented and how you are implementing `applyGamma()`. These two functions expect all the channels to be the same (or null).

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -84,6 +134,31 @@
+
+    RetainPtr<NSArray> arguments = @[
+        inputImage.get(),
+        @(m_effect->redFunction().amplitude),   @(m_effect->redFunction().exponent),   @(m_effect->redFunction().offset),
+        @(m_effect->greenFunction().amplitude), @(m_effect->greenFunction().exponent), @(m_effect->greenFunction().offset),
+        @(m_effect->blueFunction().amplitude),  @(m_effect->blueFunction().exponent),  @(m_effect->blueFunction().offset)
```
You can access `amplitude`, `exponent` and `offset` only when type of the function is `FECOMPONENTTRANSFER_TYPE_GAMMA`. See https://www.w3.org/TR/filter-effects-1/#transfer-function-element

nullhook replied:
all members of ComponentTransferFunction are always valid and initialized. i'm not sure what you mean by "access" here. it's not a union-based structure.

shallawa replied:
The spec says about the attribute `type`:
> type = "[identity](https://www.w3.org/TR/filter-effects-1/#attr-valuedef-type-identity) | [table](https://www.w3.org/TR/filter-effects-1/#attr-valuedef-type-table) | [discrete](https://www.w3.org/TR/filter-effects-1/#attr-valuedef-type-discrete) | [linear](https://www.w3.org/TR/filter-effects-1/#attr-valuedef-type-linear) | [gamma](https://www.w3.org/TR/filter-effects-1/#attr-valuedef-type-gamma)"
Indicates the type of component transfer function. The type of function determines the applicability of the other attributes.

And it says about the attribute `offset`:
> offset = `"<number>"`
When type="gamma", the offset of the gamma function.
> The [initial value](https://svgwg.org/svg2-draft/types.html#TermInitialValue) for [offset](https://www.w3.org/TR/filter-effects-1/#element-attrdef-fecomponenttransfer-offset) is 0.

My understanding from these is when `type="gamma"` the value of the attribute `offset` is in fact describing the transfer function. But when `type="linear"`, for example, the value of the attribute `offset` should not be used because it is just the initial value which is `0` and it is not part of describing the transfer function.

nullhook replied:
so, when a channel's type isn't gamma we're modeling the function with an identity transform for that channel? 

for ex, if channel G is not a gamma type, the function is `amplitude * pow(input, exponent) + offset` and parameters should be amp=1.0, exp=1.0, offset=0.0 but currently we're passing as all 0?

shouldn't this be initialized with right values in the struct here then?
```
struct ComponentTransferFunction {
    ...
    float slope { 0 };
    float intercept { 0 };
    float amplitude { 1 };
    float exponent { 1 };
    float offset { 0 };
};
```
or we do this when we're binding the arguments to the shader?

```
auto getGammaParams = [](const ComponentTransferFunction& func) {
    if (func.type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA) {
        return std::make_tuple(func.amplitude, func.exponent, func.offset);
    }
    // For non-gamma types, use identity transform
    return std::make_tuple(1.0f, 1.0f, 0.0f);
};
```

nullhook replied:
done

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +108,15 @@
+        || m_effect->blueFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA;
+
+    if (needsGamma)
+        return applyGamma(inputImage, result);
```
Since you moved the applying the gamma type to `applyGamma()`, should we move the other part to `applyLinear()`.

nullhook replied:
sure

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,56 @@
+            || function.type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+#endif
             || function.type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_LINEAR;
     };
```
I think this should stay the same `isNullOrLinear()` and we should add `isNullOrGamma()`. `FECOMPONENTTRANSFER_TYPE_LINEAR` and `FECOMPONENTTRANSFER_TYPE_GAMMA` functions cannot be applied together.

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,56 @@
+    return isSupported(effect.redFunction())
+        && isSupported(effect.greenFunction())
+        && isSupported(effect.blueFunction())
+        && isSupported(effect.alphaFunction());
```
I think this should be 
```
return isNullOrLinear(effect.redFunction()) && NullOrLinear(effect.greenFunction()) ...
    || isNullOrGamma(effect.redFunction()) && isNullOrGamma(effect.greenFunction()) ...;
```

nullhook replied:
this can get verbose especially with a directive and support for other filter types in the future

```
#if HAVE(CI_KERNEL)
    return (isNullOrLinear(effect.redFunction())  && isNullOrLinear(effect.greenFunction()) ...
        || (isNullOrGamma(effect.redFunction()) && isNullOrGamma(effect.greenFunction()) ...;
#else
    return isNullOrLinear(effect.redFunction()) && isNullOrLinear(effect.greenFunction()) ...;
```

maybe?

```
auto allChannelsMatch = [&](auto predicate) { return predicate(effect.redFunction()) && predicate(effect.greenFunction()) ...; };
    
#if HAVE(CI_KERNEL)
    return allChannelsMatch(isNullOrLinear) || allChannelsMatch(isNullOrGamma) || allChannelsMatch(isNullOrIdentity) ...;
#else
    return allChannelsMatch(isNullOrLinear);
#endif
```

shallawa commented on Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
```diff
@@ -40,10 +40,13 @@ enum class ComponentTransferType : uint8_t {
     ComponentTransferType type { ComponentTransferType::FECOMPONENTTRANSFER_TYPE_UNKNOWN };
 
+    // For linear
     float slope { 0 };
```
I think fixing the initials of these members should be in a separate PR. Maybe it is okay since the change is small. But the spec says:

slope = "[\<number\>](https://www.w3.org/TR/css3-values/#number-value)"
When type="linear", the slope of the linear function.

The [initial value](https://svgwg.org/svg2-draft/types.html#TermInitialValue) for [slope](https://www.w3.org/TR/filter-effects-1/#element-attrdef-fecomponenttransfer-slope) is 1.

shallawa replied:
Please change the initial value of this member before landing.

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,66 @@
+#if HAVE(CI_KERNEL)
+static CIKernel* gammaKernel()
+{
+    static NeverDestroyed<RetainPtr<CIKernel>> kernel;
+    static std::once_flag onceFlag;
+
+    std::call_once(onceFlag, [] {
```
I think you can use this pattern instead:
```
static NeverDestroyed<RetainPtr<CIKernel>> kernel = [] {
    ...
    return kernels[0];
}();
````

mwyrzykowski replied:
@nullhook is this called from multiple threads? If so, I think we need to use call_once otherwise the static initialization isn't thread safe

nullhook replied:
i don't think we're calling this code from multiple threads. i just wanted to make it future-proof and keep it thread-safe using call_once. however, the pattern @shallawa suggests might break here.

mwyrzykowski replied:
personally I would prefer keeping the `call_once` as-is @shallawa as static initializers are not thread safe in WebKit

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,66 @@
+            return;
+        }
+
+        kernel.get() = kernels[0];
```
Should we use `adoptNS(kernels[0])` here?

nullhook replied:
`kernel = adoptNS(kernels[0]);` 

> "Overload resolution selected deleted operator '='"

i don't think NeverDestroyed forwards assignment ops

nullhook replied:
can do the following `kernel.gets() = adoptNS(kernels[0])` but eventually crashes.

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSArrayI applyWithExtent:roiCallback:arguments:]: unrecognized selector sent to instance 0x74ad1cec0'
```

nullhook replied:
`kernel.get() = adoptNS([kernels[0] retain]);` this doesn't seem to crash. adoptNS expects an owned object?

shallawa replied:
I think it is okay to leave it without `adoptNS()` because it is a static global variable. It is going to be freed when the WebProcess is terminated.

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,66 @@
 }
 
+#if HAVE(CI_KERNEL)
+static CIKernel* gammaKernel()
```
I think this can be
```
static RetainPtr<CIKernel>& gammaKernel()
```

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -45,17 +47,66 @@
+        kernel.get() = kernels[0];
+    });
+
+    return kernel.get().get();
```
If the function returns `RetainPtr<CIKernel>&`, I think this can be just `return kernel;`

shallawa commented on Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
```diff
@@ -67,6 +118,20 @@
+    bool isGammaTransform = m_effect->redFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->greenFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->blueFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA
+        || m_effect->alphaFunction().type == ComponentTransferType::FECOMPONENTTRANSFER_TYPE_GAMMA;
```
I think this code and the the code of `supportsCoreImageRendering()` should be shared. I think this part will be clearer if it is written like this:
```
    if (allChannelsMatch(isNullOrGamma))
        return applyGamma(inputImage, result);
    return applyLinear(inputImage, result);
```

nullhook replied:
the checks are now shared.

nullhook committed:
Add GPU acceleration for gamma component transfer filters
https://bugs.webkit.org/show_bug.cgi?id=296882
rdar://problem/157477030

Reviewed by Said Abou-Hallawa and Simon Fraser.

Support gamma-type feComponentTransfer in Core Image path. Previously,
only linear transfers were hardware-accelerated, forcing gamma
transforms to use slower software rendering.

Source/WTF/wtf/PlatformHave.h:
Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.h:
Source/WebCore/platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm:
(WebCore::isNullOr):
(WebCore::allChannelsMatch):
(WebCore::gammaKernel):
(WebCore::FEComponentTransferCoreImageApplier::supportsCoreImageRendering): Deleted.
(WebCore::FEComponentTransferCoreImageApplier::apply const): Deleted.
Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:

Canonical link: https://commits.webkit.org/298602@main
